### ISSUE TITLE
Profile Heapster on large clusters

### ISSUE THREAD

**Issue by vishh** [kubernetes/kubernetes#15716]
Heapster is currently required for multiple autoscaling features in kubernetes - Horizontal pod, node and initial resource estimation. 
To meet the scalability goals for kubernetes v1.1, we need to test heapster's stability on a large cluster (250 nodes) with lots of pods (30 per node). 

Measuring compute resource overhead on a large cluster will be useful for users.

Profiling heapster's cpu and memory usage will be useful to understand the potential bottlenecks under a scale scenario. 

One thing to be aware is that the current resource `limits` on heapster are not applicable for cluster's greater than ~10 nodes.

**Comment by bgrant0607** [comment]
Anything that needs to be done for 1.1 is P0 at this point.

**Comment by dchen1107** [comment]
For some background, here is what @saad-ali and I had done for 1.0: https://github.com/kubernetes/kubernetes/issues/10335

**Comment by davidopp** [comment]
@wojtek-t

**Comment by wojtek-t** [comment]
@gmarek - since you were already doing it yesterday for smaller clusters - can you please continue it for larger ones? Thanks!

**Comment by gmarek** [comment]
Sure - I spoke with Filip yesterday, and that's what I'll be doing today.

**Comment by gmarek** [comment]
I'll paste the results later today (I need to finish experiments). I also spent some time trying to plug in pprof in a way that actually produces useful profiles, but failed.

I can connect to the pprof endpoints easily enough, but I'm getting empty (for any reasonable definition of empty) profiles. When I hit `/heap` profile I'm seeing all symbol information and quite a lot of data, so either it gets discarded at some point, or pprof can't use some of the data. I don't have this kind of problems when profiling API server or anything else.

My guess is that it's the result of not having (?) compiled heapster binary and lack of symbol data, but that's just a hunch.

**Comment by gmarek** [comment]
@vishh @bgrant0607 @fgrzadkowski @mwielgus @jszczepkowski @wojtek-t 
TL;DR: I didn't manage to run 250 Node experiments, but probable result will be that heapster won't fit on n1-standard-1 machine, as it'll use ~3GB of memory.

Details:
Single experiment consisted of:
- starting a fresh cluster and noting heapster memory usage (as reported by `docker stats`)
- running a density '30 test, monitoring memory usage and noting highest memory usage
- waiting for the test to finish (namespace deletion) and some more time (probably ~30minutes to let heapster GC do its work) and noting memory usage
- running a density '30 test again
- waiting again

I wanted to run load test on at least few clusters, but I run out of time (single experiment takes way over an hour walltime and doing them in parallel would be a pain). I'll run remaining 250 node tests and load tests on a 50/100/250 node clusters on Monday.

Yesterdays/todays results:

no. of Nodes | startup mem | 1st run mem | 1st break mem | 2nd run mem | finish mem |
------------------|------------------|-------------------|---------------------|--------------------|----------------|
10 | 25 MB | 96 MB	| 63 MB | |
20 | 55 MB |207 MB	 | 132 MB | |
30 | 71 MB | 302 MB | 167 MB | |
40 | 118 MB | 432 MB | 207 MB | |
50 | 106 MB (?) | 487 MB (?) | 263 MB | 519 MB | 313 MB
75 | 148 MB | 843 MB (?) | 415 MB | 921 MB (?) | 458 MB
100 | 369 MB | 938 MB | 571 MB | 955 MB | 619 MB
150 | 513 MB | 1.76 GB |	800 MB | 2.07 GB |

Some side observations:
- GC in Heapster is really, really slow - in cluster with a big churn (e.g. one in which MapReduce kind of workload) it's unlikely that it'll be able to keep up.
- It does not seem to leak resources, I left 40 Node cluster overnight and Heapster behaved correctly
- It should be clear from the data, that usage is more or less linear, so I don't expect any surprises for 250 Node cluster (as long as Heapster won't be CPU bounded, but I haven't seen much CPU usage)

If someone have any idea how to get `pprof` working I'll be happy to add some profiles to this data.

**Comment by fgrzadkowski** [comment]
@gmarek Thanks much for running this experiment. This is very useful data.

I think this shows that we can easily use current heapster in 1.1 release.

The only AI that I see left before 1.1 is to write down the recommended limits somewhere in user guides. Once that's done I think we should move this issue out of 1.1 and assign P1 or P2 priority.

**Comment by derekwaynecarr** [comment]
@kubernetes/rh-cluster-infra @DirectXMan12 @smarterclayton - see above data on Heapster memory consumption relative to number of nodes and 30 pods per node.

**Comment by vishh** [comment]
@gmarek: Thanks for the data! This information is very useful! 

Since you mentioned you are having issues profiling heapster, here is what I usually do for profiling:
1. I download the release binary from Github. Since the current release tagged with Kubernetes v1.1 is v0.18.2, download the corresponding heapster binary:
``` wget https://github.com/kubernetes/heapster/releases/download/v0.18.2/heapster```

2. Start a local kubectl proxy and access heapster pprof endpoint through the proxy:
```go tool pprof heapster http://localhost:8001/api/v1/proxy/namespaces/kube-system/services/heapster/debug/pprof/heap```

This usually works for me. Let me know if you run into any issues.

**Comment by vishh** [comment]
> GC in Heapster is really, really slow - in cluster with a big churn (e.g. one in which MapReduce kind of workload) it's unlikely that it'll be able to keep up.

The default retention policy for dead containers is now 4 minutes. We can tune that down further or even not keep dead containers around. AFAIK the only use case that might need this data is GKE where scrape heapster every 2 minutes. But we could argue that missing the last sample isn't that bad. cc @mwielgus 
Switching from lists to static slices for caching might also help here. List requires Go's GC to bring down the memory usage. In the past I have noticed that the go runtime keeps nearly 30% memory around that is not in use, possibly based on heuristics. 

> It does not seem to leak resources, I left 40 Node cluster overnight and Heapster behaved correctly

Glad to know this. 

> It should be clear from the data, that usage is more or less linear, so I don't expect any surprises for 250 Node cluster (as long as Heapster won't be CPU bounded, but I haven't seen much CPU usage)

Heapster's update frequency has been reduced to 30seconds recently by @piosz. That might be the reason for reduced cpu usage. If we ever increase that frequency, we should measure impact on CPU.

**Comment by vishh** [comment]
> I think this shows that we can easily use current heapster in 1.1 release.

@fgrzadkowski: +1

**Comment by gmarek** [comment]
@vishh I'll try what you suggested. I was accessing heapster profile through API server proxy.

**Comment by gmarek** [comment]
@vishh - thanks, adding the binary solved problems with empty profiles. If anyone else want to profile it starting up a proxy is (probably) one way, another one is setting up an ssh tunnel to the master machine:
```
ssl kubernetes-master -L1234:localhost:8080
```
and using pprof locally:
```
go tool pprof heapster localhost:1234/api/v1/proxy/namespaces/kube-system/services/heapster/debug/pprof/<profile>
```

The advantage of the latter is that it doesn't require setting up a kubeproxy and will allow profiling of pretty much anything (as long as it's proxy-able through master, unsecured port exists and it exposes pprof endpoint)

**Comment by piosz** [comment]
@gmarek @vishh how about adding a short paragraph to [debugging](https://github.com/kubernetes/heapster/blob/master/docs/debugging.md) doc?

**Comment by gmarek** [comment]
For 250 Node cluster there're no surprises:

before 1st density | 1st density | 1st break | 2nd density 
-------------------------|---------------|--------------|-----------------
591 MB | 2.77 GB | 1.1 GB | >3GB (OOM)

I also run load test on 100 Node cluster. It was using 1.02 GB of memory in the peak.
Profile from load run:
[heapster100.pdf](https://github.com/kubernetes/kubernetes/files/15947/heapster100.pdf)
Profile from idle cluster after the run:
[heapster_after.pdf](https://github.com/kubernetes/kubernetes/files/15948/heapster_after.pdf)

**Comment by smarterclayton** [comment]
Seems like we are allocating massive amounts of temporary objects, so while
we aren't leaking or holding on to anything excessive, we're not very
efficient in our allocations (likely similar to the apiserver profile you
had earlier).  Can you grab the allocation space/count profile?

On Oct 19, 2015, at 7:50 AM, Marek Grabowski <notifications@github.com>
wrote:

For 250 Node cluster there're no surprises:
before 1st density 1st density 1st break 2nd density 591 MB 2.77 GB 1.1 GB >3GB
(OOM)

I also run load test on 100 Node cluster. It was using 1.02 GB of memory in
the peak.
Profile from load run:
heapster100.pdf
<https://github.com/kubernetes/kubernetes/files/15947/heapster100.pdf>
Profile from idle cluster after the run:
heapster_after.pdf
<https://github.com/kubernetes/kubernetes/files/15948/heapster_after.pdf>

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/15716#issuecomment-149192350>
.

**Comment by gmarek** [comment]
@smarterclayton I'm not sure if this is what you wanted:
alloc-space:
[heapster-alloc-space.pdf](https://github.com/kubernetes/kubernetes/files/15985/heapster-alloc-space.pdf)
alloc-objects:
[heapster-alloc-objects.pdf](https://github.com/kubernetes/kubernetes/files/15986/heapster-alloc-objects.pdf)

**Comment by smarterclayton** [comment]
Half of all allocations being from float -> string seems high - if this is
on a write path (write to response, write to disk) strconv.AppendFloat
would allow us to reuse byte buffers.  Need to dig a bit more.

**Comment by gmarek** [comment]
If some is interested - after a night Heapster memory usage on an idle 100 Node cluster was 460 MB

**Comment by davidopp** [comment]
Does this mean #13048 should be considered a blocker for 1,1? We are planning to advertise that Kubernetes scales up to 250 nodes; that statement has to include the cluster addons that are integral to its operation (like Heapster). heapster-controller.yaml sets a memory limit of 300M so it seems either we need to significantly reduce Heapster memory consumption or implement #13048. (It would probably be sufficient to just do the latter since n1-standard-2 VM is 7.5G) Am I misunderstanding?

**Comment by gmarek** [comment]
As long as we don't officially support cluster autoscaling it should be enough to compute requirements at startup, which should be simpler.

**Comment by davidopp** [comment]
Agreed - so, item (1) from my comment here
https://github.com/kubernetes/kubernetes/issues/13048#issuecomment-133564022

Who knows enough about the cluster setup scripts to dynamically generate heapster-controller.yaml based on the cluster size?

(BTW I'm not saying we shouldn't *also* try to reduce Heapster memory consumption -- clearly we should. But it seems we're so far from 300M that we're going to need to do some kind of dynamic sizing regardless.)

**Comment by wojtek-t** [comment]
>
Agreed - so, item (1) from my comment here
#13048 (comment)

+1

**Comment by fgrzadkowski** [comment]
@davidopp I don't think that automatic requests for Heapster should be a blocker for 1.1. We had exactly the same problem for 1.1 and it's not easy to do properly, especially with manual/automatic cluster scaling. We talked we should just document recommended limits in user-guides and don't block the release.

**Comment by fgrzadkowski** [comment]
Downgrading priority as we agreed current heapster is good enough for 1.1

**Comment by wojtek-t** [comment]
I think that definitely should NOT focus on the case of automatic/manual cluster scaling and just document it. However, setting limits based on the initial cluster is quite important, because we know that it will be OOMing if we set it to small.

**Comment by fgrzadkowski** [comment]
@wojtek-t It is important, but it's not a new problem. Heapster had exactly the same issue in 1.0 (actually it was much worse earlier). HPA is a beta feature, so I think it's ok to have some manual (well documented) actions that have to be performed.

**Comment by wojtek-t** [comment]
I think it depends how difficult it is - if this is just few lines of code, we should do it, otherwise probably documenting it is fine.

**Comment by vishh** [comment]
Applying better initial limits based on cluster size will definitely help users. In the past many users did not read our documentation and instead were battling OOM kills. 
Will HPA be enabled in GKE?

**Comment by davidopp** [comment]
Moving back to 1.1 for now, pending additional discussion.

**Comment by davidopp** [comment]
@alex-mohr

**Comment by fgrzadkowski** [comment]
I don't understand why we discuss this in the context of HPA. It is a beta feature, so it does not have to scale to 100 or 250 nodes.

I'm not arguing that it will not improve UX - for sure it would be better to have it. I'm just arguing that it is not a blocker for 1.1. As said before HPA is a beta feature so it does not have to scale in the same way as the rest of the cluster. For heapster itself we had exactly the same problem earlier (or to be precise it was much worse in 1.0). So why is it a blocker?

Initially we created this issue as a P0 just to make sure that with the current default limits we can work in most cases. @gmarek verified that we can (we run with ~900 pods)

IMHO we tend to push far too many things to 1.1 which delays the release.

**Comment by davidopp** [comment]
We are not considering this issue part of the HPA feature. My understanding from Marek's measurements is that Heapster will OOM even without HPA enabled, on large clusters, since it was consuming multiple GB of memory just running density test without any HPA. Please correct me if I am misunderstanding. (We understand that this problem was also in 1.0, but now that we are advertising 250 node clusters it seems people are more likely to try larger and hit this problem. It is especially problematic on GKE where there is no way for user to modify their heapster-controller.yaml)

@brendandburns volunteered to implement dynamic generation of limit in heapster-controller.yaml so I am assigning this issue to him and marking it 1.1. Thanks Brendan!!!

**Comment by piosz** [comment]
@davidopp, will dynamic generation happen only once the cluster is created or also during its resize?

**Comment by davidopp** [comment]
Only at cluster creation time. It does not solve the full problem but we don't think many people dynamically resize their cluster. Full fix is punted to 1.2, issue #13048.

**Comment by wojtek-t** [comment]
>
Only at cluster creation time. It does not solve the full problem but we don't think many people dynamically resize their cluster. Full fix is punted to 1.2, issue #13048.

+1

**Comment by dchen1107** [comment]
I am closing this issue since #16185 was merged.
